src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdater.java

```diff
@@ -36,6 +39,8 @@
 
   private Notification newModulesVersionsNotification = null;
 
+  private final List<String> notifiedModules = new ArrayList<>();
```

Comment:
Should this be
1. a (hash)set?
2. synced / "concurrent" collection?

Reply:
If you had `Set<String> notifiedModules = ConcurrentHashMap.newKeySet()`, your filtering method could be just:
```java
return modules.stream()
  .filter(m -> notifiedModules.add(m.getName())
  .collect(Collectors.toList());
```
because `Set.add(elem)` returns true iff `elem` wasn't there in the first place.

Yes, side effects in stream filters are bad but this is one of those rare situations where we accept them.

Reply:
That's better so I did that

Reply:
I get shivers just looking at that filter :smile: 

But more importantly, I see no reason to use a set with concurrency support here. Just a standard `HashSet` will do

Reply:
Argh, there is of course the case where `restart` is called, and it is then possible that the `run` method of a single updater instance is executing on two different threads simultaneously, so it should be synced after all. Sorry, my bad.

But that has me thinking now, is it even safe for `run` to be running on two threads at the same time?

@OlliKiljunen I remember we had some `IllegalStateException` stack traces from `IntelliJCourse#unregister`, but I can't find them anymore. `run` calls `unregister` for the course stored in `MainViewModel`, so this could actually be the root cause of the illegal state where unregister is called multiple times.

Reply:
I'll take a look at that.

> I get shivers just looking at that filter 😄

Those are good shivers. :+1: But it's important to notice that the side effects here are _idempotent_ which makes it less scary.

Reply:
I think it might make more sense to make `run` synchronized instead of using a synced set here.